executor/joiner.go

```diff
@@ -176,108 +181,111 @@ type semiJoiner struct {
 	}
 
 	for inner := inners.Current(); inner != inners.End(); inner = inners.Next() {
 		j.makeShallowJoinRow(j.outerIsRight, inner, outer)
 
-		matched, err = expression.EvalBool(j.ctx, j.conditions, j.shallowRow.ToRow())
+		matched, _, err = expression.EvalBool(j.ctx, j.conditions, j.shallowRow.ToRow())
 		if err != nil {
-			return false, errors.Trace(err)
+			return false, false, errors.Trace(err)
 		}
 		if matched {
 			chk.AppendPartialRow(0, outer)
 			inners.ReachEnd()
-			return true, nil
+			return true, false, nil
 		}
 	}
-	return false, nil
+	return false, false, nil
 }
 
-func (j *semiJoiner) onMissMatch(outer chunk.Row, chk *chunk.Chunk) {
+func (j *semiJoiner) onMissMatch(hasNull bool, outer chunk.Row, chk *chunk.Chunk) {
 }
 
 type antiSemiJoiner struct {
 	baseJoiner
 }
 
 // tryToMatch implements joiner interface.
-func (j *antiSemiJoiner) tryToMatch(outer chunk.Row, inners chunk.Iterator, chk *chunk.Chunk) (matched bool, err error) {
+func (j *antiSemiJoiner) tryToMatch(outer chunk.Row, inners chunk.Iterator, chk *chunk.Chunk) (matched bool, hasNull bool, err error) {
 	if inners.Len() == 0 {
-		return false, nil
+		return false, false, nil
 	}
 
 	if len(j.conditions) == 0 {
 		inners.ReachEnd()
-		return true, nil
+		return true, false, nil
 	}
 
 	for inner := inners.Current(); inner != inners.End(); inner = inners.Next() {
 		j.makeShallowJoinRow(j.outerIsRight, inner, outer)
 
-		matched, err = expression.EvalBool(j.ctx, j.conditions, j.shallowRow.ToRow())
+		matched, isNull, err := expression.EvalBool(j.ctx, j.conditions, j.shallowRow.ToRow())
 		if err != nil {
-			return false, errors.Trace(err)
+			return false, false, errors.Trace(err)
 		}
 		if matched {
 			inners.ReachEnd()
-			return true, nil
+			return true, false, nil
 		}
+		hasNull = hasNull || isNull
 	}
-	return false, nil
+	return false, hasNull, nil
 }
 
-func (j *antiSemiJoiner) onMissMatch(outer chunk.Row, chk *chunk.Chunk) {
-	chk.AppendRow(outer)
+func (j *antiSemiJoiner) onMissMatch(hasNull bool, outer chunk.Row, chk *chunk.Chunk) {
+	if !hasNull {
+		chk.AppendRow(outer)
+	}
 }
 
 type leftOuterSemiJoiner struct {
 	baseJoiner
 }
 
 // tryToMatch implements joiner interface.
-func (j *leftOuterSemiJoiner) tryToMatch(outer chunk.Row, inners chunk.Iterator, chk *chunk.Chunk) (matched bool, err error) {
+func (j *leftOuterSemiJoiner) tryToMatch(outer chunk.Row, inners chunk.Iterator, chk *chunk.Chunk) (matched bool, hasNull bool, err error) {
 	if inners.Len() == 0 {
-		return false, nil
+		return false, false, nil
 	}
 
 	if len(j.conditions) == 0 {
 		j.onMatch(outer, chk)
 		inners.ReachEnd()
-		return true, nil
+		return true, false, nil
 	}
 
 	for inner := inners.Current(); inner != inners.End(); inner = inners.Next() {
 		j.makeShallowJoinRow(false, inner, outer)
 
-		matched, err = expression.EvalBool(j.ctx, j.conditions, j.shallowRow.ToRow())
+		matched, _, err = expression.EvalBool(j.ctx, j.conditions, j.shallowRow.ToRow())
```

Comment:
For "left outer semi join", I think we should make the return value `hasNull` be correct. Because in this join method, what we really needed is the join result: matched or is NULL. For example, if we don't consider `hasNull`, we won't be able to produce the correct result for this query:

```sql
TiDB(root@127.0.0.1:test) > desc select a in (select b from t) from t;
+-------------------------+-------+------+----------------------------------------------------------------------------+
| id                      | count | task | operator info                                                              |
+-------------------------+-------+------+----------------------------------------------------------------------------+
| Projection_6            | 4.00  | root | 5_aux_0                                                                    |
| └─HashLeftJoin_7        | 4.00  | root | left outer semi join, inner:TableReader_11, equal:[eq(test.t.a, test.t.b)] |
|   ├─TableReader_9       | 4.00  | root | data:TableScan_8                                                           |
|   │ └─TableScan_8       | 4.00  | cop  | table:t, range:[-inf,+inf], keep order:false, stats:pseudo                 |
|   └─TableReader_11      | 4.00  | root | data:TableScan_10                                                          |
|     └─TableScan_10      | 4.00  | cop  | table:t, range:[-inf,+inf], keep order:false, stats:pseudo                 |
+-------------------------+-------+------+----------------------------------------------------------------------------+
6 rows in set (0.00 sec)

TiDB(root@127.0.0.1:test) > select a in (select b from t) from t;
+------------------------+
| a in (select b from t) |
+------------------------+
|                      1 |
|                      0 |
|                      0 |
|                      0 |
+------------------------+
4 rows in set (0.00 sec)
```

the table was generated by:
```sql
create table t(a bigint, b bigint, c bigint);
insert into t values(1, 1, 1), (null, 2, 2), (3, null, 3), (4, 400, 400);
```

the correct result is:
```sql
MySQL(root@localhost:test) > select a in (select b from t) from t;
+------------------------+
| a in (select b from t) |
+------------------------+
|                      1 |
|                   NULL |
|                   NULL |
|                   NULL |
+------------------------+
4 rows in set (0.00 sec)
```

Reply:
This should be done after we differentiate whether the join key is from `in` or not, for example, here is the output from MySQL:
```
mysql> select a in (select b from t) from t;
+------------------------+
| a in (select b from t) |
+------------------------+
|                      1 |
|                   NULL |
|                   NULL |
|                   NULL |
+------------------------+
4 rows in set (0.00 sec)

mysql> select exists(select * from t t2 where t2.b = t1.a) from t t1;
+----------------------------------------------+
| exists(select * from t t2 where t2.b = t1.a) |
+----------------------------------------------+
|                                            1 |
|                                            0 |
|                                            0 |
|                                            0 |
+----------------------------------------------+
4 rows in set (0.00 sec)

mysql> select a in (select a from t t2 where t2.b = t1.a) from t t1;
+---------------------------------------------+
| a in (select a from t t2 where t2.b = t1.a) |
+---------------------------------------------+
|                                           1 |
|                                           0 |
|                                           0 |
|                                           0 |
+---------------------------------------------+
4 rows in set (0.00 sec)
```